components/layout_2020/flow/float.rs

```diff
@@ -70,6 +71,112 @@ impl ContainingBlockPositionInfo {
     }
 }
 
+/// This data strucure is used to try to place non-floating content among float content.
+/// This is used primarily to place replaced content and independent formatting contexts
+/// next to floats, as the specifcation dictates.
+pub(crate) struct PlacementAmongFloats<'a> {
+    /// The [FloatContext] to use for this placement.
+    float_context: &'a FloatContext,
+    /// The current bands we are considering for this placement.
+    current_bands: VecDeque<FloatBand>,
+    /// The size of the object to place.
+    object_size: Vec2<Length>,
+    /// The minimum position in the block direction for the placement. Objects should not
+    /// be placed before this point.
+    ceiling: Length,
+}
+
+impl<'a> PlacementAmongFloats<'a> {
+    pub(crate) fn new(
+        float_context: &'a FloatContext,
+        ceiling: Length,
+        object_size: Vec2<Length>,
+    ) -> Self {
+        let current_bands = VecDeque::from([float_context.bands.find(ceiling).unwrap()]);
+        PlacementAmongFloats {
+            float_context,
+            current_bands,
+            object_size,
+            ceiling,
+        }
+    }
+
+    fn top_of_placement_for_current_bands(&self) -> Length {
+        self.ceiling.max(self.current_bands.front().unwrap().top)
+    }
+
+    fn current_bands_height(&self) -> Length {
+        assert!(self.current_bands.len() > 0);
+        self.current_bands.back().unwrap().top - self.top_of_placement_for_current_bands()
```

Comment:
If flooring by `self.ceiling` makes a difference, then this height could be negative?

Reply:
Hrm...yeah, I guess we don't want to find the first band that this object fits in, but the first band that extends beyond the ceiling. This might need a bit of work.

Reply:
On second thought, I think this is correct. There are two options here:

1. The ceiling is in the middle of the first band in which case the the usable height of our band selection goes from [ceiling...end of bands].
2. The first band starts lower than the ceiling in which case the usable height of our band selection goes from [start of bands...end of bands].

Reply:
Mm, yeah I guess that's right, the problem is in `calculate_viable_inline_space`.